<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introductory tutorial for R Package nbTransmission • nbTransmission</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introductory tutorial for R Package nbTransmission">
<meta property="og:description" content="nbTransmission">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nbTransmission</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/adjusted-odds-ratio-example-vignette.html">Adjusted Odds Ratio Example</a>
    </li>
    <li>
      <a href="../articles/nbTransmission-vignette.html">Introductory tutorial for R Package nbTransmission</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sarahleavitt/nbTransmission/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introductory tutorial for R Package
nbTransmission</h1>
                        <h4 data-toc-skip class="author">Sarah V.
Leavitt</h4>
            
            <h4 data-toc-skip class="date">2025-05-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sarahleavitt/nbTransmission/blob/HEAD/vignettes/nbTransmission-vignette.Rmd" class="external-link"><code>vignettes/nbTransmission-vignette.Rmd</code></a></small>
      <div class="hidden name"><code>nbTransmission-vignette.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This will walk you through how to use the package nbTransmission to
analyze an infectious disease outbreak. You want to start by installing
nbTransmission if you have not already done so using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'sarahleavitt/nbTransmission'</span><span class="op">)</span></span></code></pre></div>
<p>Then you can load the package by running:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://sarahleavitt.github.io/nbTransmission/" class="external-link">'nbTransmission'</a></span><span class="op">)</span></span></code></pre></div>
<p>If you want the results to match this tutorial, set your seed to
match:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>This package implements an algorithm that takes a dataset of ordered
possible infector-infectee pairs from an infectious disease outbreak or
outbreak cluster and estimates the relative probability the cases are
linked by direct transmission using a classification technique called
naive Bayes (NB).</p>
<p><strong>Naive Bayes</strong> is a simple and popular machine learning
algorithm that uses Bayes rule to estimate the probability of an outcome
in a prediction dataset given a set of covariates from the observed
frequencies in a training dataset. It is a two step process:</p>
<ol style="list-style-type: decimal">
<li>Determine the association between the outcome and the covariates in
a training dataset where the outcome is known.</li>
<li>Estimate the probability of the outcome given the covariates in a
prediction dataset where the outcome is not known.</li>
</ol>
<p>In this context, a subset of cases with pathogen whole genome
sequence (WGS) or contact investigation data serve as the training
dataset and the outcome is if the pair is “linked” by one or both of
these metrics which is a proxy for being linked by direct transmission.
The covariates can be any categorical variables and could represent
spatial, clinical, demographic, and temporal characteristics of the
cases. The goal is to use these covariates to predict the probability
that pairs are linked even when we do not have any WGS or contact
data.</p>
<p>Because the outcomes in the training set represent probable and not
certain transmission events and a given case could have mulitple
probable infectors, we use an iterative estimation procedure. This
procedure randomly chooses one link of all of the possible links to
include in the training dataset multiple times, and then uses mxn cross
prediction to give all pairs a turn in the prediction dataset.</p>
</div>
<div class="section level2">
<h2 id="creating-a-dataset-of-pairs">Creating a Dataset of Pairs<a class="anchor" aria-label="anchor" href="#creating-a-dataset-of-pairs"></a>
</h2>
<p>Lets say you have a dataset of individuals from an outbreak. We are
going to use a simulated dataset called indData included in the package.
This dataset has 100 individuals and contains an ID column, a column
with the true infector (because this is simulated data so we know the
truth), the infection date and time, the sampling date and time, and
four covariates (X1-X4). The covariates are artificial but were designed
to resemble some possible covariates in real applications.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">indData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">indData</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    100 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ individualID : num  1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;  $ infector     : num  0 1 2 2 2 3 5 5 6 7 ...</span></span>
<span><span class="co">#&gt;  $ infectionDate: POSIXct, format: "1986-04-08 20:31:37" "1986-09-11 04:55:40" ...</span></span>
<span><span class="co">#&gt;  $ sampleDate   : POSIXct, format: "1987-08-03 09:10:04" "1989-05-22 08:46:29" ...</span></span>
<span><span class="co">#&gt;  $ X1           : chr  "a" "a" "a" "b" ...</span></span>
<span><span class="co">#&gt;  $ X2           : chr  "a" "a" "a" "b" ...</span></span>
<span><span class="co">#&gt;  $ X3           : chr  "a" "b" "b" "a" ...</span></span>
<span><span class="co">#&gt;  $ X4           : chr  "e" "e" "e" "d" ...</span></span></code></pre></div>
<p>Now we need to create a dataset of all ordered pairs of cases to
analyze. The nbTranmssion package includes a function called indToPair
which will do exactly this. The function takes an individual-level
dataset and creates a pair-level dataset. You tell the function the name
of the ID column and the resulting dataset will have an pair ID column
that combines the two individual IDs. It will also include all other
columns with the suffixes “.1” and “.2” corresponding to the two
individuals in the pair. If you specify a date column it will also find
the time difference between the pairs and if you set ordered to TRUE, it
will only include pairs where the date for individual 1 is before
individual 2. Here is how you would use this function and what the
resulting data frame looks like:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairDataRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/indToPair.html">indToPair</a></span><span class="op">(</span><span class="va">indData</span>,</span>
<span>                         indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                         separator <span class="op">=</span> <span class="st">"_"</span>,</span>
<span>                         dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>                         units <span class="op">=</span> <span class="st">"days"</span>,</span>
<span>                         ordered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pairDataRaw</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    9900 obs. of  18 variables:</span></span>
<span><span class="co">#&gt;  $ individualID.2    : num  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ individualID.1    : num  47 75 84 28 65 57 49 83 62 31 ...</span></span>
<span><span class="co">#&gt;  $ pairID            : chr  "47_1" "75_1" "84_1" "28_1" ...</span></span>
<span><span class="co">#&gt;  $ infector.1        : num  32 104 79 21 53 37 32 77 43 22 ...</span></span>
<span><span class="co">#&gt;  $ infectionDate.1   : POSIXct, format: "1994-10-20 06:41:42" "1999-02-19 09:11:04" ...</span></span>
<span><span class="co">#&gt;  $ sampleDate.1      : POSIXct, format: "1995-01-26 03:24:26" "1999-04-03 08:10:08" ...</span></span>
<span><span class="co">#&gt;  $ X1.1              : chr  "a" "b" "a" "b" ...</span></span>
<span><span class="co">#&gt;  $ X2.1              : chr  "b" "a" "c" "a" ...</span></span>
<span><span class="co">#&gt;  $ X3.1              : chr  "b" "a" "a" "b" ...</span></span>
<span><span class="co">#&gt;  $ X4.1              : chr  "d" "j" "c" "d" ...</span></span>
<span><span class="co">#&gt;  $ infector.2        : num  0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;  $ infectionDate.2   : POSIXct, format: "1986-04-08 20:31:37" "1986-04-08 20:31:37" ...</span></span>
<span><span class="co">#&gt;  $ sampleDate.2      : POSIXct, format: "1987-08-03 09:10:04" "1987-08-03 09:10:04" ...</span></span>
<span><span class="co">#&gt;  $ X1.2              : chr  "a" "a" "a" "a" ...</span></span>
<span><span class="co">#&gt;  $ X2.2              : chr  "a" "a" "a" "a" ...</span></span>
<span><span class="co">#&gt;  $ X3.2              : chr  "a" "a" "a" "a" ...</span></span>
<span><span class="co">#&gt;  $ X4.2              : chr  "e" "e" "e" "e" ...</span></span>
<span><span class="co">#&gt;  $ infectionDate.Diff: num  -3116 -4700 -4379 -3199 -3489 ...</span></span></code></pre></div>
<p>The next step is to create pair-level equivalents of the covariates.
This can be done in any way that is clinically meaningful. For example
if the individual-level covariate was nationality, the pair-level
covariate could be do the individuals have the same nationality or not.
If the individual-level covariate was town of residence, the pair-level
covariate could be do the individuals live in the same town, neighboring
towns, or further away. Currently the method only supports pair-level
covariates that are categorical and they need to also be factor
variables.</p>
<p>Pair-level covariates for this simulated outbreak are already stored
in the data frame pairData which is also included in this package and
this is the dataset we are going to analyze.</p>
<p>The pairData dataset also has a column called snpDist with the number
of single nucleotide polymorphisms (SNPs) separating the pathogen of the
two individuals, a categorical representation of the difference between
infection times, as well as a column called transmission indicating if
the pair is a true transmission pair (because again this is simulated
data so we know this!).</p>
<p>Here are details about the individual-level covariates and how they
were transformed into pair-level covariates. They were simulated to be
associated with transmission in different degrees.</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="29%">
<col width="20%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th align="center">Variable</th>
<th align="center">Individual-level</th>
<th align="center">Pair-level</th>
<th align="center">Motivation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">X1/Z1</td>
<td align="center">2 options: a, b</td>
<td align="center">Z1 = 1 if same <br> Z1 = 0 if different</td>
<td align="center">Sex</td>
</tr>
<tr class="even">
<td align="center">X2/Z2</td>
<td align="center">4 options: a, b, c, d</td>
<td align="center">Z2 = 0 if same <br> Z2 = 0 if different</td>
<td align="center">Nationality</td>
</tr>
<tr class="odd">
<td align="center">X3/Z3</td>
<td align="center">2 options: a, b</td>
<td align="center">Z3 = 1 if a-a <br> Z3 = 2 if b-b <br> Z3 = 3 if a-b
<br> Z3 = 4 if b-a</td>
<td align="center">Homelessness</td>
</tr>
<tr class="even">
<td align="center">X4/Z4</td>
<td align="center">10 options: a-j</td>
<td align="center">Z4 = 1 if same <br> Z4 = 2 if adjacent <br> Z4 = 3 if
other</td>
<td align="center">County of residence</td>
</tr>
<tr class="odd">
<td align="center">timeCat</td>
<td align="center">infectionDate</td>
<td align="center">Time between cases <br> timeCat = 1 if &lt;1yr <br>
timeCat = 2 if 1-2yrs <br> timeCat = 3 if 2-3yrs <br> timeCat = 4 if
3-4yrs. <br> timeCat = 5 if 4-5yrs. <br> timeCat = 6 if &gt;5yrs.</td>
<td align="center">Generation interval</td>
</tr>
</tbody>
</table>
<p>Here are the distributions of the individual-level covariates:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">indData</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    a    b </span></span>
<span><span class="co">#&gt; 0.48 0.52</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">indData</span><span class="op">$</span><span class="va">X2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    a    b    c    d </span></span>
<span><span class="co">#&gt; 0.48 0.24 0.24 0.04</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">indData</span><span class="op">$</span><span class="va">X3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    a    b </span></span>
<span><span class="co">#&gt; 0.67 0.33</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">indData</span><span class="op">$</span><span class="va">X4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    a    c    d    e    f    g    h    j </span></span>
<span><span class="co">#&gt; 0.01 0.12 0.42 0.32 0.09 0.02 0.01 0.01</span></span></code></pre></div>
<p>And the pair-level covariates by whether or not the pair is a true
transmission pair (TRUE) or not (FALSE)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairData</span><span class="op">$</span><span class="va">transmission</span>, <span class="va">pairData</span><span class="op">$</span><span class="va">Z1</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        </span></span>
<span><span class="co">#&gt;                 0         1</span></span>
<span><span class="co">#&gt;   FALSE 0.5050443 0.4949557</span></span>
<span><span class="co">#&gt;   TRUE  0.4137931 0.5862069</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairData</span><span class="op">$</span><span class="va">transmission</span>, <span class="va">pairData</span><span class="op">$</span><span class="va">Z2</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        </span></span>
<span><span class="co">#&gt;                 0         1</span></span>
<span><span class="co">#&gt;   FALSE 0.6626923 0.3373077</span></span>
<span><span class="co">#&gt;   TRUE  0.2873563 0.7126437</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairData</span><span class="op">$</span><span class="va">transmission</span>, <span class="va">pairData</span><span class="op">$</span><span class="va">Z3</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        </span></span>
<span><span class="co">#&gt;                 1         2         3         4</span></span>
<span><span class="co">#&gt;   FALSE 0.4469581 0.1060838 0.2236829 0.2232752</span></span>
<span><span class="co">#&gt;   TRUE  0.4137931 0.1724138 0.1839080 0.2298851</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairData</span><span class="op">$</span><span class="va">transmission</span>, <span class="va">pairData</span><span class="op">$</span><span class="va">Z4</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        </span></span>
<span><span class="co">#&gt;                  1          2          3</span></span>
<span><span class="co">#&gt;   FALSE 0.29175583 0.43646184 0.27178233</span></span>
<span><span class="co">#&gt;   TRUE  0.65517241 0.33333333 0.01149425</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="estimating-relative-transmission-probabilties">Estimating Relative Transmission Probabilties<a class="anchor" aria-label="anchor" href="#estimating-relative-transmission-probabilties"></a>
</h2>
<p>There are two more data preparation steps that we need to do before
we are ready to find the relative transmission probabilities. First we
need subset the pair dataset to only ordered pairs (if you are using the
indToPair function, you can use the ordered = TRUE option and skip this
step!)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairDataOrdered</span> <span class="op">&lt;-</span> <span class="va">pairData</span><span class="op">[</span><span class="va">pairData</span><span class="op">$</span><span class="va">infectionDate.2</span> <span class="op">&gt;=</span> <span class="va">pairData</span><span class="op">$</span><span class="va">infectionDate.1</span>, <span class="op">]</span></span></code></pre></div>
<p>Next we need to create the variable that defines probable links for
our training dataset. We will use the SNP distance to define probable
links. We are going to say that pairs with fewer than 3 SNPs will be
considered probable links and pairs with more than 10 SNPs are
considered non-linked. Any pairs with 3-10 SNPs will be considered
indeterminate and not included in the training set. Those pairs as well
as any pairs without WGS info will only be use in the prediction
dataset. The variable that defines the probable links should be a
logical with TRUE indicating a link and FALSE inidicating a
non-link.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairDataOrdered</span><span class="op">$</span><span class="va">snpClose</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">pairDataOrdered</span><span class="op">$</span><span class="va">snpDist</span> <span class="op">&lt;</span> <span class="fl">3</span>, <span class="cn">TRUE</span>,</span>
<span>                                   <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">pairDataOrdered</span><span class="op">$</span><span class="va">snpDist</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="cn">FALSE</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairDataOrdered</span><span class="op">$</span><span class="va">snpClose</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE  &lt;NA&gt; </span></span>
<span><span class="co">#&gt;  1467   247  3236</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pairDataOrdered</span><span class="op">$</span><span class="va">snpClose</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      FALSE       TRUE       &lt;NA&gt; </span></span>
<span><span class="co">#&gt; 0.29636364 0.04989899 0.65373737</span></span></code></pre></div>
<p>Now we can use the function nbTransmission to calculate the relative
transmission probabilities for all pairs, training the model using the
WGS data. For the argument, indIDVar, the function looks for that value
with the suffix “.1” and “.2” in the pair data frame (which would have
been automatically created if you used indToPair to create the
pair-level data frame). The options n, m, and nReps indicate how many
times the iterative procedure. This procedure randomly chooses one link
of all of the possible links to include in the training dataset nReps
times, and then uses mxn cross prediction for each training dataset to
give all pairs a turn in the prediction dataset. Therefore in total you
run naive Bayes nReps * n * m times. <strong>In this tutorial, nReps = 1
for computation speed but nReps should be at least 10.</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resGen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nbProbabilities.html">nbProbabilities</a></span><span class="op">(</span>orderedPair <span class="op">=</span> <span class="va">pairDataOrdered</span>,</span>
<span>                            indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                            pairIDVar <span class="op">=</span> <span class="st">"pairID"</span>,</span>
<span>                            goldStdVar <span class="op">=</span> <span class="st">"snpClose"</span>,</span>
<span>                            covariates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Z1"</span>, <span class="st">"Z2"</span>, <span class="st">"Z3"</span>, <span class="st">"Z4"</span>, <span class="st">"timeCat"</span><span class="op">)</span>,</span>
<span>                            label <span class="op">=</span> <span class="st">"SNPs"</span>, l <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                            n <span class="op">=</span> <span class="fl">10</span>, m <span class="op">=</span> <span class="fl">1</span>, nReps <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">resGen</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ probabilities:'data.frame':   4950 obs. of  7 variables:</span></span>
<span><span class="co">#&gt;   ..$ label     : chr [1:4950] "SNPs" "SNPs" "SNPs" "SNPs" ...</span></span>
<span><span class="co">#&gt;   ..$ pairID    : chr [1:4950] "001_002" "022_003" "020_003" "011_003" ...</span></span>
<span><span class="co">#&gt;   ..$ pAvg      : num [1:4950] 0.9083 0.0512 0.0437 0.0428 0.0404 ...</span></span>
<span><span class="co">#&gt;   ..$ pSD       : num [1:4950] 0.28983 0.00508 0.00768 0.00674 0.00658 ...</span></span>
<span><span class="co">#&gt;   ..$ pScaled   : num [1:4950] 1 0.158 0.135 0.132 0.125 ...</span></span>
<span><span class="co">#&gt;   ..$ pRank     : num [1:4950] 1 1 2 3 4 5 6 6 8 9 ...</span></span>
<span><span class="co">#&gt;   ..$ nEstimates: int [1:4950] 10 10 10 10 10 10 10 10 10 10 ...</span></span>
<span><span class="co">#&gt;  $ estimates    :'data.frame':   17 obs. of  7 variables:</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:17] "SNPs" "SNPs" "SNPs" "SNPs" ...</span></span>
<span><span class="co">#&gt;   ..$ level    : chr [1:17] "Z1:0" "Z1:1" "Z2:0" "Z2:1" ...</span></span>
<span><span class="co">#&gt;   ..$ nIter    : int [1:17] 10 10 10 10 10 10 10 10 10 10 ...</span></span>
<span><span class="co">#&gt;   ..$ logorMean: num [1:17] 0 -0.0585 0 0.818 0 ...</span></span>
<span><span class="co">#&gt;   ..$ logorSE  : num [1:17] NaN 0.259 NaN 0.254 NaN ...</span></span>
<span><span class="co">#&gt;   ..$ logorCILB: num [1:17] NaN -0.566 NaN 0.321 NaN ...</span></span>
<span><span class="co">#&gt;   ..$ logorCIUB: num [1:17] NaN 0.449 NaN 1.316 NaN ...</span></span></code></pre></div>
<p>The results are a list with two data frames, the first of which
containing the average and scaled relative transmission probabilities.
We can extract that data frame and merge it back with our pair-level
dataset for further analysis. We can see that the algorithm assigned
much higher probabilities to the true transmission pairs than the rest
of the pairs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nbResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">resGen</span><span class="op">$</span><span class="va">probabilities</span>, <span class="va">pairDataOrdered</span>, by <span class="op">=</span> <span class="st">"pairID"</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">nbResults</span><span class="op">$</span><span class="va">pScaled</span>, <span class="va">nbResults</span><span class="op">$</span><span class="va">transmission</span>, <span class="va">summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`FALSE`</span></span>
<span><span class="co">#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">#&gt; 0.0000394 0.0014305 0.0041901 0.0166358 0.0118997 0.8714328 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`TRUE`</span></span>
<span><span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">#&gt; 0.004128 0.029727 0.054669 0.208047 0.411729 1.000000</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">nbResults</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pScaled</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">transmission</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/results-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="covariates-associated-with-probable-transmission">Covariates Associated with Probable Transmission<a class="anchor" aria-label="anchor" href="#covariates-associated-with-probable-transmission"></a>
</h2>
<p>The second data frame contains log odds ratios with 95% confidence
intervals describing the contribution of each covariate value to the
probabilities. These log odds ratios represent the association between
the covariates and how training links were defined but modified by the
iterative estimation procedure which corrects the training dataset to
more resemble the dataset of true links and nonlinks (for example by
only selecting one possible infector for each case).</p>
<p>The log odds ratios calculated for each covariate value at each
iteration of the naive Bayes transmission algorithm and the estimates in
the data frame are averaged across the iterations. The confidence
intervals are estimated using Rubin’s rules, a method developed for
multiple imputation, to summarize the error across the iterations.</p>
<p>Log odds ratios that are greater than 0 mean that having that value
of that covariate would result in an increase in the estimated
transmission probability for that pair compared to having the reference
value. Conversely log odds ratios that are less than 0 mean that having
that value of that covariate would result in a decrease in the estimated
transmission probability for that paire compared to having the reference
value.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span><span class="co">#Exponentiating the log odds ratios and creating a table of odds ratios</span></span>
<span><span class="va">orTab</span> <span class="op">&lt;-</span> <span class="va">resGen</span><span class="op">$</span><span class="va">estimates</span></span>
<span><span class="va">orTab</span><span class="op">$</span><span class="va">orMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">orTab</span><span class="op">$</span><span class="va">logorMean</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">orTab</span><span class="op">$</span><span class="va">orCILB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">orTab</span><span class="op">$</span><span class="va">logorCILB</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">orTab</span><span class="op">$</span><span class="va">orCIUB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">orTab</span><span class="op">$</span><span class="va">logorCIUB</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">orTab</span><span class="op">$</span><span class="va">Value</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"[A-z0-9]+\\:"</span>, <span class="st">""</span>, <span class="va">orTab</span><span class="op">$</span><span class="va">level</span><span class="op">)</span></span>
<span><span class="va">orTab</span><span class="op">$</span><span class="va">Variable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\:[A-z0-9+-&lt;=&gt;]+"</span>, <span class="st">""</span>, <span class="va">orTab</span><span class="op">$</span><span class="va">level</span><span class="op">)</span></span>
<span><span class="va">orTab</span> <span class="op">&lt;-</span> <span class="va">orTab</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Variable"</span>, <span class="st">"Value"</span>, <span class="st">"orMean"</span>, <span class="st">"orCILB"</span>, <span class="st">"orCIUB"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div class="section level4">
<h4 id="table-of-contribution-of-covariates">Table of contribution of covariates<a class="anchor" aria-label="anchor" href="#table-of-contribution-of-covariates"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Variable</th>
<th align="left">Value</th>
<th align="right">orMean</th>
<th align="right">orCILB</th>
<th align="right">orCIUB</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">7</td>
<td align="left">Z1</td>
<td align="left">0</td>
<td align="right">1.00</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">Z1</td>
<td align="left">1</td>
<td align="right">0.94</td>
<td align="right">0.57</td>
<td align="right">1.57</td>
</tr>
<tr class="odd">
<td align="left">9</td>
<td align="left">Z2</td>
<td align="left">0</td>
<td align="right">1.00</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
</tr>
<tr class="even">
<td align="left">10</td>
<td align="left">Z2</td>
<td align="left">1</td>
<td align="right">2.27</td>
<td align="right">1.38</td>
<td align="right">3.73</td>
</tr>
<tr class="odd">
<td align="left">11</td>
<td align="left">Z3</td>
<td align="left">1</td>
<td align="right">1.00</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
</tr>
<tr class="even">
<td align="left">12</td>
<td align="left">Z3</td>
<td align="left">2</td>
<td align="right">1.22</td>
<td align="right">0.54</td>
<td align="right">2.77</td>
</tr>
<tr class="odd">
<td align="left">13</td>
<td align="left">Z3</td>
<td align="left">3</td>
<td align="right">1.37</td>
<td align="right">0.72</td>
<td align="right">2.59</td>
</tr>
<tr class="even">
<td align="left">14</td>
<td align="left">Z3</td>
<td align="left">4</td>
<td align="right">1.20</td>
<td align="right">0.64</td>
<td align="right">2.24</td>
</tr>
<tr class="odd">
<td align="left">15</td>
<td align="left">Z4</td>
<td align="left">1</td>
<td align="right">1.00</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
</tr>
<tr class="even">
<td align="left">16</td>
<td align="left">Z4</td>
<td align="left">2</td>
<td align="right">0.56</td>
<td align="right">0.34</td>
<td align="right">0.94</td>
</tr>
<tr class="odd">
<td align="left">17</td>
<td align="left">Z4</td>
<td align="left">3</td>
<td align="right">0.18</td>
<td align="right">0.08</td>
<td align="right">0.44</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">timeCat</td>
<td align="left">1</td>
<td align="right">1.00</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
</tr>
<tr class="odd">
<td align="left">2</td>
<td align="left">timeCat</td>
<td align="left">2</td>
<td align="right">0.88</td>
<td align="right">0.46</td>
<td align="right">1.68</td>
</tr>
<tr class="even">
<td align="left">3</td>
<td align="left">timeCat</td>
<td align="left">3</td>
<td align="right">0.64</td>
<td align="right">0.29</td>
<td align="right">1.41</td>
</tr>
<tr class="odd">
<td align="left">4</td>
<td align="left">timeCat</td>
<td align="left">4</td>
<td align="right">0.41</td>
<td align="right">0.15</td>
<td align="right">1.11</td>
</tr>
<tr class="even">
<td align="left">5</td>
<td align="left">timeCat</td>
<td align="left">5</td>
<td align="right">0.12</td>
<td align="right">0.03</td>
<td align="right">0.59</td>
</tr>
<tr class="odd">
<td align="left">6</td>
<td align="left">timeCat</td>
<td align="left">6</td>
<td align="right">0.21</td>
<td align="right">0.10</td>
<td align="right">0.44</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="forest-plot-of-contribution-of-covariates">Forest plot of contribution of covariates<a class="anchor" aria-label="anchor" href="#forest-plot-of-contribution-of-covariates"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">orTab</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Value</span>, y <span class="op">=</span> <span class="va">orMean</span>, ymin <span class="op">=</span> <span class="va">orCILB</span>,</span>
<span>                           ymax <span class="op">=</span> <span class="va">orCIUB</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Variable</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Odds ratio with 95% confidence interval"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        strip.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0</span>, vjust <span class="op">=</span> <span class="fl">1</span>, angle <span class="op">=</span> <span class="fl">360</span><span class="op">)</span>,</span>
<span>        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">0</span>, l <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">1</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/covarFig-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="finding-the-most-likely-infectors">Finding the Most Likely Infectors<a class="anchor" aria-label="anchor" href="#finding-the-most-likely-infectors"></a>
</h2>
<p>In order to find the most likely infectors, you want to find which
cases have a cluster of infectors with probabilities that are higher
than the rest of the possible infectors. To find these cases and the top
cluster of probable infectors, you can use the function clusterInfectors
which implements various clustering methods.</p>
<p>The methods you can choose from include: 1) simply taking the top n
infectors, 2) hierarchical clustering to split the infectors and the
size of the gap between clusters to choose which cases have high
probability infectors, or 3) kernel density estimation to determine if
the probabilities of all possible infectors represents one or two
distinct densities. The function also requires a cutoff value whose
meaning depends on the clustering method.</p>
<p>First let’s use hierarchical clustering to split the infectors into
two clusters and require the gap between the probabilities in each group
to be greater than 0.05.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Clustering the probabilities</span></span>
<span><span class="va">clustHC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterInfectors.html">clusterInfectors</a></span><span class="op">(</span><span class="va">nbResults</span>, indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>                           clustMethod <span class="op">=</span> <span class="st">"hc_absolute"</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clustHC</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    1    2 </span></span>
<span><span class="co">#&gt;   89 4861</span></span>
<span><span class="co">#Subsetting to just the top cluster</span></span>
<span><span class="va">topClustHC</span> <span class="op">&lt;-</span> <span class="va">clustHC</span><span class="op">[</span><span class="va">clustHC</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span></span></code></pre></div>
<p>With this clustering method, the 4950 possible pairs are reduced to
89 most likely infectors. These pairs represent 83 of the 99 cases. The
remaining cases do not have any clear top cluster of infectors. The plot
bellow shows an example of a case that has a clear top cluster of
infectors with this clustering method (case 32 on the left) and one that
does not (case 40 on the right).</p>
<div class="section level4">
<h4 id="clustering-illustration">Clustering Illustration<a class="anchor" aria-label="anchor" href="#clustering-illustration"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">clustHC</span><span class="op">[</span><span class="va">clustHC</span><span class="op">$</span><span class="va">individualID.2</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">32</span>, <span class="fl">40</span><span class="op">)</span>, <span class="op">]</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pRank</span>, y <span class="op">=</span> <span class="va">pScaled</span>, color <span class="op">=</span> <span class="va">cluster</span>, shape <span class="op">=</span> <span class="va">transmission</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">individualID.2</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/clustExample-1.png" width="672"></p>
<p>An alternative method would be using kernel density estimation. With
this method, the cutoff specified is the binwidth, which we will set to
0.01 for this example.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Clustering the probabilities</span></span>
<span><span class="va">clustKD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterInfectors.html">clusterInfectors</a></span><span class="op">(</span><span class="va">nbResults</span>, indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>                           clustMethod <span class="op">=</span> <span class="st">"kd"</span>, cutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clustKD</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    1    2 </span></span>
<span><span class="co">#&gt;  103 4847</span></span>
<span><span class="co">#Subsetting to just the top cluster</span></span>
<span><span class="va">topClustKD</span> <span class="op">&lt;-</span> <span class="va">clustKD</span><span class="op">[</span><span class="va">clustKD</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span></span></code></pre></div>
<p>With this clustering method, the 4950 possible pairs are reduced to
103 most likely infectors. These pairs represent 87 of the 99 cases. The
remaining cases do not have any clear top cluster of infectors.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-results">Visualizing Results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h2>
<p>The nbTransmission package comes with two functions to visualize the
results of the algorithm. The first one, nbNetwork, draws a network of
all of the cases connected by either all possible edges (if clustMethod
= “none”) or only the top cluster of infectors as defined by the
clustering method and cutoff. The darker the edge, the higher the
probability that the pair is a transmission link.</p>
<div class="section level4">
<h4 id="network-with-all-pairs">Network with all pairs<a class="anchor" aria-label="anchor" href="#network-with-all-pairs"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0.2</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nbNetwork.html">nbNetwork</a></span><span class="op">(</span><span class="va">nbResults</span>, indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>pVar <span class="op">=</span> <span class="st">"pScaled"</span>, clustMethod <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/networkFull-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="network-with-only-top-probability-pairs">Network with only top probability pairs<a class="anchor" aria-label="anchor" href="#network-with-only-top-probability-pairs"></a>
</h4>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0.2</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nbNetwork.html">nbNetwork</a></span><span class="op">(</span><span class="va">nbResults</span>, indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>pVar <span class="op">=</span> <span class="st">"pScaled"</span>, clustMethod <span class="op">=</span> <span class="st">"hc_absolute"</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/networkTop-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="heatmap-with-stars-on-top-cluster">Heatmap with stars on top cluster<a class="anchor" aria-label="anchor" href="#heatmap-with-stars-on-top-cluster"></a>
</h4>
<p>The second function, nbHeatmap, creates a heatmap of the transmission
probabilities. The rows are the possible infectors and the columns are
the possible infectees both ordered by date of observation. The darker
the square the higher the probability that the pair represented by that
square is a transmission link. If a cluster method and cutoff are
specified, then stars will be drawn in the squares of the infectors in
the top cluster. <em>Note that this plot is better for small outbreaks
and takes a long time to draw and may not look good if the number of
cases in the outbreak is more than 200.</em></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nbHeatmap.html">nbHeatmap</a></span><span class="op">(</span><span class="va">nbResults</span>, indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>          pVar <span class="op">=</span> <span class="st">"pScaled"</span>, clustMethod <span class="op">=</span> <span class="st">"hc_absolute"</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/heatmap-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-the-reproductive-number">Estimating the Reproductive Number<a class="anchor" aria-label="anchor" href="#estimating-the-reproductive-number"></a>
</h2>
<p>One way to use the relative transmission probabilities is to estimate
the reproductive number. This package includes a function estimateR
which estimates the individual-level (<span class="math inline">\(R_i\)</span>), time-level (<span class="math inline">\(R_t\)</span>), and overall average effective
reproductive (<span class="math inline">\(\hat{R}_t\)</span>) numbers.
The function requires various variable name inputs as well as the time
frame for the <span class="math inline">\(R_t\)</span>. Currently
supported time frames are “days”, “weeks”, “months”, and “years”.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rInitial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateR.html">estimateR</a></span><span class="op">(</span><span class="va">nbResults</span>,</span>
<span>                      dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>                      indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                      pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>                      timeFrame <span class="op">=</span> <span class="st">"months"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Please choose the stable portion of the outbreak to calculate the average Rt</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">rInitial</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ RiDf       :'data.frame': 99 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ individualID : num [1:99] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   ..$ infectionDate: POSIXct[1:99], format: "1986-04-08 20:31:37" "1986-09-11 04:55:40" ...</span></span>
<span><span class="co">#&gt;   ..$ Ri           : num [1:99] 5.708 2.917 0.934 3.216 1.539 ...</span></span>
<span><span class="co">#&gt;   ..$ nInfectees   : int [1:99] 99 98 84 95 97 73 92 96 55 77 ...</span></span>
<span><span class="co">#&gt;  $ RtDf       :'data.frame': 73 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;   ..$ time    : chr [1:73] "1986-04" "1986-09" "1986-10" "1986-12" ...</span></span>
<span><span class="co">#&gt;   ..$ timeRank: int [1:73] 1 6 7 9 11 12 16 21 25 29 ...</span></span>
<span><span class="co">#&gt;   ..$ Rt      : num [1:73] 5.708 2.917 1.539 0.802 3.216 ...</span></span>
<span><span class="co">#&gt;  $ RtAvgDf    :'data.frame': 1 obs. of  1 variable:</span></span>
<span><span class="co">#&gt;   ..$ RtAvg: num 1.01</span></span>
<span><span class="co">#&gt;  $ timeFrame  : chr "months"</span></span>
<span><span class="co">#&gt;  $ rangeForAvg: NULL</span></span></code></pre></div>
<p>Note the message that is printed when running this funciton without
specifying rangeForAvg. In order to get a more accurate estimate of the
average effective reproductive number, you only want to use the stable
portion of the outbreak. You plot the <span class="math inline">\(R_t\)</span> values over time using the plotRt
function to find this portion of the outbreak. In this case I want
average the values between month 25 and month 125 which is excluding
about the first 10% and last 20% of the outbreak.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRt.html">plotRt</a></span><span class="op">(</span><span class="va">rInitial</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/cutting-1.png" width="576"></p>
<p>Then I can rerun the estimateR function specifying the range for the
average and also asking for bootstrap confidence intervals. <strong>In
this tutorial, bootSamples = 10 for computation speed but bootSamples
should be at least 100.</strong></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rFinal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateR.html">estimateR</a></span><span class="op">(</span><span class="va">nbResults</span>, dateVar <span class="op">=</span> <span class="st">"infectionDate"</span>,</span>
<span>             indIDVar <span class="op">=</span> <span class="st">"individualID"</span>, pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>             timeFrame <span class="op">=</span> <span class="st">"months"</span>, rangeForAvg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">125</span><span class="op">)</span>,</span>
<span>             bootSamples <span class="op">=</span> <span class="fl">10</span>, alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rFinal</span><span class="op">$</span><span class="va">RtAvgDf</span></span>
<span><span class="co">#&gt;      RtAvg  ciLower  ciUpper</span></span>
<span><span class="co">#&gt; 1 1.103004 0.898509 1.439759</span></span></code></pre></div>
<p>Finally, you can plot the final results adding the confidence
intervals and average <span class="math inline">\(R_t\)</span> estimates
using the plotRt function.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRt.html">plotRt</a></span><span class="op">(</span><span class="va">rFinal</span>, includeRtAvg <span class="op">=</span> <span class="cn">TRUE</span>, includeRtCI <span class="op">=</span> <span class="cn">TRUE</span>, includeRtAvgCI <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/unnamed-chunk-2-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="estimating-the-serial-interval">Estimating the Serial Interval<a class="anchor" aria-label="anchor" href="#estimating-the-serial-interval"></a>
</h2>
<p>The relative transmission probabilities can also be used to estimate
the serial interval distribution using the PEM algorithm developed by
Hens et al. 2012, extending their method to include restricting analysis
to the top cluster of possible infectors. The method can be performed
with any serial interval distribution, but this version of the package
assumes that the serial interval has a gamma distribution. The function
does allow for a shifted gamma distribution where any observed serial
intervals that are less than this shift will have probability 0. This
parameter should be used if there is a clinically lower bound for the
possible serial interval.</p>
<p>The function requires the user to indicate the column that specifies
the difference in time between the two cases in each pair. To be a true
estimate of the serial interval this variable should represent the time
between symptom onset. The units of the estimated gamma distribution
will be defined by the units of the provided. The value of the shift
should be in the same units.</p>
<p>All pairs of cases can be used in the estimation process by setting
clustMethod = “none”. However, if the probabilities are from a algorithm
such as nbProbabilities (as illustrated in this tutorial), then it is
recommeneded to use a clustering method and only include the top cluster
of infectors for infectees which have such a cluster. Another important
consideration is the time difference between cases should not be
included in the probability estimation as this will skew the
results.</p>
<p>First we will re-estimate the probabilites, excluding the timeCat
variable which was derived from the time difference between cases.
<strong>Again, nReps should be at least 10.</strong></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resGenNoT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nbProbabilities.html">nbProbabilities</a></span><span class="op">(</span>orderedPair <span class="op">=</span> <span class="va">pairDataOrdered</span>,</span>
<span>                            indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                            pairIDVar <span class="op">=</span> <span class="st">"pairID"</span>,</span>
<span>                            goldStdVar <span class="op">=</span> <span class="st">"snpClose"</span>,</span>
<span>                            covariates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Z1"</span>, <span class="st">"Z2"</span>, <span class="st">"Z3"</span>, <span class="st">"Z4"</span><span class="op">)</span>,</span>
<span>                            label <span class="op">=</span> <span class="st">"SNPs"</span>, l <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                            n <span class="op">=</span> <span class="fl">10</span>, m <span class="op">=</span> <span class="fl">1</span>, nReps <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">nbResultsNoT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">resGenNoT</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">pairDataOrdered</span>, by <span class="op">=</span> <span class="st">"pairID"</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can estimate the serial interval distribution using the top
cluster of infectors from our hierarchical clustering method. The units
for this serial interval will be in years since that is the units of the
time difference variable and will be for a standard gamma distribution
because no shift is specified.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">siPars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateSI.html">estimateSI</a></span><span class="op">(</span><span class="va">nbResultsNoT</span>,</span>
<span>                     indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                     timeDiffVar <span class="op">=</span> <span class="st">"infectionDiffY"</span>,</span>
<span>                     pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>                     clustMethod <span class="op">=</span> <span class="st">"hc_absolute"</span>,</span>
<span>                     cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                     initialPars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">siPars</span></span>
<span><span class="co">#&gt;   clustMethod cutoff nIndividuals nInfectors  pCluster    shape    scale</span></span>
<span><span class="co">#&gt; 1 hc_absolute   0.05           77   1.012987 0.7777778 1.079997 1.907707</span></span>
<span><span class="co">#&gt;     meanSI medianSI     sdSI</span></span>
<span><span class="co">#&gt; 1 2.060318 1.470459 1.982545</span></span></code></pre></div>
<p>I can also run the estimateSI function asking for bootstrap
confidence intervals. The confidence interval calculation can take a
while. <strong>In this tutorial, bootSamples = 5 for computation speed
but bootSamples should be at least 100.</strong></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">siParsCI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateSI.html">estimateSI</a></span><span class="op">(</span><span class="va">nbResultsNoT</span>,</span>
<span>                       indIDVar <span class="op">=</span> <span class="st">"individualID"</span>,</span>
<span>                       timeDiffVar <span class="op">=</span> <span class="st">"infectionDiffY"</span>,</span>
<span>                       pVar <span class="op">=</span> <span class="st">"pScaled"</span>,</span>
<span>                       clustMethod <span class="op">=</span> <span class="st">"hc_absolute"</span>,</span>
<span>                       cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                       initialPars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                       bootSamples <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">siParsCI</span></span>
<span><span class="co">#&gt;   clustMethod cutoff nIndividuals nInfectors  pCluster    shape    scale</span></span>
<span><span class="co">#&gt; 1 hc_absolute   0.05           77   1.012987 0.7777778 1.079997 1.907707</span></span>
<span><span class="co">#&gt;     meanSI medianSI     sdSI shapeCILB shapeCIUB scaleCILB scaleCIUB meanCILB</span></span>
<span><span class="co">#&gt; 1 2.060318 1.470459 1.982545 0.8086973  1.080874  1.932677  2.516003 2.015324</span></span>
<span><span class="co">#&gt;   meanCIUB medianCILB medianCIUB   sdCILB   sdCIUB</span></span>
<span><span class="co">#&gt; 1 2.395427   1.386426   1.689443 1.977227 2.464418</span></span></code></pre></div>
<p>The serial interval can be plotted over the histogram of the true
pairs (because we are using simulated data)</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">truePairs</span> <span class="op">&lt;-</span> <span class="va">nbResultsNoT</span><span class="op">[</span><span class="va">nbResultsNoT</span><span class="op">$</span><span class="va">transmission</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">nbResultsNoT</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">infectionDiffY</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">truePairs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">40</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Serial Interval (years)"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">dgamma</a></span><span class="op">(</span><span class="va">infectionDiffY</span>, shape <span class="op">=</span> <span class="va">siPars</span><span class="op">$</span><span class="va">shape</span>, scale <span class="op">=</span> <span class="va">siPars</span><span class="op">$</span><span class="va">scale</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="nbTransmission-vignette_files/figure-html/siPlot-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this tutorial, we have illustrated how to use the nbTransmission
package to do the following:</p>
<ul>
<li>Create a dataset of pairs from a dataset of individuals</li>
<li>Estimate the relative transmission probabilities for all pairs of
cases in an outbreak</li>
<li>Find the contribution of the covariate values to the
probabilities</li>
<li>Find the top cluster of infectors for each case</li>
<li>Estimate the effective reproductive number</li>
<li>Estimate the serial interval distribution</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Sarah V Leavitt, Anne Shapiro.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
